File: mkdocs/snippets/devbook/chain/prove_mosaic_definition.py
import hashlib
import json
import os
import urllib.request
from binascii import unhexlify

from symbolchain.BufferWriter import BufferWriter
from symbolchain.CryptoTypes import Hash256
from symbolchain.symbol.Merkle import (
	deserialize_patricia_tree_nodes,
	prove_patricia_merkle
)

NODE_URL = os.getenv('NODE_URL', 'https://reference.symboltest.net:3001')
print(f'Using node {NODE_URL}')

try:
	# Fetch the network currency mosaic ID [>step-1]
	url = f'{NODE_URL}/network/properties'
	with urllib.request.urlopen(url) as response:
		props = json.loads(response.read().decode())
	raw_id = props['chain']['currencyMosaicId']
	mosaic_id = int(raw_id.replace("'", ""), 16)
	mosaic_id_hex = f'{mosaic_id:016X}'
	print(f'Currency mosaic ID: {mosaic_id_hex}')

	# Fetch the mosaic properties
	mosaic_path = f'/mosaics/{mosaic_id_hex}'
	print(f'Fetching mosaic from {mosaic_path}')
	url = f'{NODE_URL}{mosaic_path}'
	with urllib.request.urlopen(url) as response:
		mosaic_data = json.loads(response.read().decode())
	mosaic = mosaic_data['mosaic']
	print(json.dumps(mosaic, indent=2))
	# [<step-1]
	# Serialize and hash the mosaic properties [>step-2]
	writer = BufferWriter()
	writer.write_int(int(mosaic['version']), 2)
	writer.write_int(int(mosaic['id'], 16), 8)
	writer.write_int(int(mosaic['supply']), 8)
	writer.write_int(int(mosaic['startHeight']), 8)
	writer.write_bytes(unhexlify(mosaic['ownerAddress']))
	writer.write_int(int(mosaic['revision']), 4)
	writer.write_int(int(mosaic['flags']), 1)
	writer.write_int(int(mosaic['divisibility']), 1)
	writer.write_int(int(mosaic['duration']), 8)
	hashed_value = Hash256(hashlib.sha3_256(writer.buffer).digest())
	print(f'Hashed value: {hashed_value}')

	# Hash the mosaic ID to get the encoded key
	writer = BufferWriter()
	writer.write_int(int(mosaic['id'], 16), 8)
	encoded_key = Hash256(hashlib.sha3_256(writer.buffer).digest())
	print(f'Encoded key: {encoded_key}')
	# [<step-2]
	# Fetch the current network height [>step-3]
	url = f'{NODE_URL}/chain/info'
	with urllib.request.urlopen(url) as response:
		chain_info = json.loads(response.read().decode())
	height = int(chain_info['height'])
	print(f'Current height: {height}')

	# Fetch the block's state hash and roots
	block_path = f'/blocks/{height}'
	print(f'Fetching block from {block_path}')
	url = f'{NODE_URL}{block_path}'
	with urllib.request.urlopen(url) as response:
		block_data = json.loads(response.read().decode())
	state_hash = Hash256(block_data['block']['stateHash'])
	sub_cache_key = 'stateHashSubCacheMerkleRoots'
	roots = [Hash256(r) for r in block_data['meta'][sub_cache_key]]
	print(f'State hash: {state_hash}')
	# [<step-3]
	# Fetch the patricia tree path [>step-4]
	tree_url = f'/mosaics/{mosaic_id_hex}/merkle'
	print(f'Fetching tree path from {tree_url}')
	url = f'{NODE_URL}{tree_url}'
	with urllib.request.urlopen(url) as response:
		tree_data = json.loads(response.read().decode())
	merkle_path = deserialize_patricia_tree_nodes(
		unhexlify(tree_data['raw']))
	print(f'Tree path: {len(merkle_path)} nodes')
	key_hex = str(encoded_key)
	key_pos = 0
	for i, node in enumerate(merkle_path):
		key_pos += node.path.size
		path_str = (f'  path: {node.hex_path}' if node.path.size else '')
		if hasattr(node, 'value'):
			print(f'  [{i}] leaf{path_str}  value: {node.value}')
		else:
			nibble = key_hex[key_pos]
			key_pos += 1
			active = [
				f'{j:X}' for j, link in enumerate(node.links) if link
			]
			print(f'  [{i}] branch{path_str}'
				f'  links: [{",".join(active)}]'
				f'  -> follow {nibble}')
	# [<step-4]
	# Verify the mosaic state [>step-5]
	result = prove_patricia_merkle(
		encoded_key, hashed_value, merkle_path, state_hash, roots)

	if result.name == 'VALID_POSITIVE':
		print(f'Mosaic {mosaic_id_hex} state verified at height {height}')
	else:
		raise RuntimeError(
			f'Mosaic {mosaic_id_hex} proof failed: {result.name}')
	# [<step-5]
except Exception as e:
	print(e)

File: mkdocs/snippets/devbook/chain/prove_mosaic_definition.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.1

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.symbol.Merkle;
import org.symbol.sdk.symbol.Merkle.BranchNode;
import org.symbol.sdk.symbol.Merkle.LeafNode;
import org.symbol.sdk.symbol.Merkle.PatriciaMerkleProofResult;
import org.symbol.sdk.symbol.Merkle.TreeNode;
import org.symbol.sdk.utils.Converter;
import org.symbol.sdk.utils.Transforms;

final class ProveMosaicDefinition {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private final String nodeUrl = System.getenv().getOrDefault(
		"NODE_URL", "https://reference.symboltest.net:3001");

	public static void main(final String[] args) {
		try {
			new ProveMosaicDefinition().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run() throws IOException, InterruptedException {
		System.out.printf("Using node %s%n", nodeUrl);

		// Fetch the network currency mosaic ID [>step-1]
		final JsonNode props = getJson("/network/properties");
		final String rawId = props.get("chain")
			.get("currencyMosaicId").asText();
		final long mosaicId = Long.parseUnsignedLong(
			rawId.substring(2).replace("'", ""), 16);
		final String mosaicIdHex = "%016X".formatted(mosaicId);
		System.out.printf("Currency mosaic ID: %s%n", mosaicIdHex);

		// Fetch the mosaic properties
		final String mosaicPath = "/mosaics/" + mosaicIdHex;
		System.out.printf("Fetching mosaic from %s%n", mosaicPath);
		final JsonNode mosaicData = getJson(mosaicPath);
		final JsonNode mosaic = mosaicData.get("mosaic");
		System.out.println(JSON_MAPPER.writerWithDefaultPrettyPrinter()
			.writeValueAsString(mosaic));
		// [<step-1]
		// Serialize and hash the mosaic properties [>step-2]
		final ByteBuffer buffer = ByteBuffer.allocate(40 + 24)
			.order(ByteOrder.LITTLE_ENDIAN);
		buffer.putShort((short) mosaic.get("version").asInt());
		buffer.putLong(Long.parseUnsignedLong(
			mosaic.get("id").asText(), 16));
		buffer.putLong(Long.parseUnsignedLong(
			mosaic.get("supply").asText()));
		buffer.putLong(Long.parseUnsignedLong(
			mosaic.get("startHeight").asText()));
		buffer.put(Converter.hexToUint8(
			mosaic.get("ownerAddress").asText()));
		buffer.putInt(mosaic.get("revision").asInt());
		buffer.put((byte) mosaic.get("flags").asInt());
		buffer.put((byte) mosaic.get("divisibility").asInt());
		buffer.putLong(Long.parseUnsignedLong(
			mosaic.get("duration").asText()));
		final CryptoTypes.Hash256 hashedValue =
			new CryptoTypes.Hash256(Transforms.sha3_256(
				buffer.array()));
		System.out.printf("Hashed value: %s%n", hashedValue);

		// Hash the mosaic ID to get the encoded key
		final ByteBuffer keyBuffer = ByteBuffer.allocate(8)
			.order(ByteOrder.LITTLE_ENDIAN);
		keyBuffer.putLong(Long.parseUnsignedLong(
			mosaic.get("id").asText(), 16));
		final CryptoTypes.Hash256 encodedKey =
			new CryptoTypes.Hash256(Transforms.sha3_256(
				keyBuffer.array()));
		System.out.printf("Encoded key: %s%n", encodedKey);
		// [<step-2]
		// Fetch the current network height [>step-3]
		final JsonNode chainInfo = getJson("/chain/info");
		final long height = chainInfo.get("height").asLong();
		System.out.printf("Current height: %d%n", height);

		// Fetch the block's state hash and roots
		final String blockPath = "/blocks/" + height;
		System.out.printf("Fetching block from %s%n", blockPath);
		final JsonNode blockData = getJson(blockPath);
		final CryptoTypes.Hash256 stateHash =
			new CryptoTypes.Hash256(blockData.get("block")
				.get("stateHash").asText());
		final List<CryptoTypes.Hash256> roots = new ArrayList<>();
		for (final JsonNode root : blockData.get("meta")
			.get("stateHashSubCacheMerkleRoots"))
			roots.add(new CryptoTypes.Hash256(root.asText()));
		System.out.printf("State hash: %s%n", stateHash);
		// [<step-3]
		// Fetch the patricia tree path [>step-4]
		final String treeUrl = "/mosaics/" + mosaicIdHex + "/merkle";
		System.out.printf("Fetching tree path from %s%n", treeUrl);
		final JsonNode treeData = getJson(treeUrl);
		final List<TreeNode> merklePath =
			Merkle.deserializePatriciaTreeNodes(
				Converter.hexToUint8(treeData.get("raw").asText()));
		System.out.printf("Tree path: %d nodes%n", merklePath.size());
		final String keyHex = encodedKey.toString();
		int keyPos = 0;
		for (int i = 0; merklePath.size() > i; ++i) {
			final TreeNode node = merklePath.get(i);
			keyPos += node.path.size();
			final String pathStr = 0 != node.path.size()
				? "  path: " + node.hexPath()
				: "";
			if (node instanceof LeafNode leaf) {
				System.out.printf("  [%d] leaf%s  value: %s%n",
					i, pathStr, leaf.value);
			} else if (node instanceof BranchNode branch) {
				final char nibble = keyHex.charAt(keyPos);
				++keyPos;
				final List<String> active = new ArrayList<>();
				for (int j = 0; branch.links.length > j; ++j) {
					if (null != branch.links[j])
						active.add("%X".formatted(j));
				}
				System.out.printf(
					"  [%d] branch%s  links: [%s]  -> follow %c%n",
					i, pathStr, String.join(",", active), nibble);
			}
		}
		// [<step-4]
		// Verify the mosaic state [>step-5]
		final int result = Merkle.provePatriciaMerkle(
			encodedKey, hashedValue, merklePath, stateHash, roots);

		if (PatriciaMerkleProofResult.VALID_POSITIVE == result) {
			System.out.printf(
				"Mosaic %s state verified at height %d%n",
				mosaicIdHex, height);
		} else {
			throw new IllegalStateException(String.format(
				"Mosaic %s proof failed: %d",
				mosaicIdHex, result));
		} // [<step-5]
	}

	private JsonNode getJson(
		final String path
	) throws IOException, InterruptedException {
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(nodeUrl + path)).GET().build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		return JSON_MAPPER.readTree(response.body());
	}
}
